torch/csrc/CudaIPCTypes.cpp

```diff
@@ -0,0 +1,51 @@
+#include <torch/csrc/CudaIPCTypes.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+CudaIPCReceivedData::CudaIPCReceivedData(std::shared_ptr<void> shared_ptr)
+    : shared_ptr_(std::move(shared_ptr)) {}
+
+int64_t CudaIPCSentData::get() {
+  return *(int64_t*)(shared_storage_ptr_.get());
+}
+
+CudaIPCSentDataLimbo::~CudaIPCSentDataLimbo() {
+  collect();
+  if (shared_blocks_.size() > 0) {
+    AT_ERROR(
+        "Producer process has been terminated before all shared CUDA tensors released.");
+  }
+}
+
+void CudaIPCSentDataLimbo::collect() {
+  std::vector<CudaIPCSentData*> kept_blocks;
+  for (auto const& sd : shared_blocks_) {
+    if (sd->get() > 0) {
+      kept_blocks.push_back(sd);
+    } else {
+      delete sd;
+    }
+  }
```

Comment:
I'm wondering if it would be good to keep track of how long `shared_blocks_` is, and report a warning if it gets very large, since in that situation you're going to end up with some sort of quadratic slowdown. We don't have very good facility for reporting warnings (glog, maybe) but it seems like it would save someone a lot of perf debugging trouble down the line.